### ISSUE TITLE
Add extract_access_token option to OAuth2::Client

### ISSUE THREAD

**Issue by pboling** [oauth-xx/oauth2#582]
Already implemented in `1-4-stable` branch, and needs to be upstreamed to `master` for release of v2.
https://github.com/oauth-xx/oauth2/pull/518/

**Comment by stanhu** [comment]
@pboling I looked at this as well. It seems to me that `extract_access_token` should not be necessary now that `access_token_class` is available. 

Anyone who wants custom behavior can override `AccessToken#from_hash` and be done with it. Can we just close this issue? 😄

**Comment by pboling** [comment]
@stanhu It certainly could be, but I want to try to provide a smooth upgrade path, so the implementations aren't fragmented more than they will already be from the jump to v2.  I think this should be relatively easy to add though.

**Comment by stanhu** [comment]
@pboling I spent time trying to resolve the conflicts from cherry-picking #518 into `master`. However, I found that supporting both would be a bit messy because the logic for extracting access tokens is exploding in a couple of directions:

1. `AccessToken#build_access_token` in `1-4-stable` does this  hack where it tries to deprecate the use of `from_hash`: https://github.com/oauth-xx/oauth2/pull/518/files#diff-62e5865bb4ee1a24c64039310c8fa6cdc5af9372d118f936c46b53231db65544R245-R249
2. However, `access_token_class` needs `from_hash` for this reason.
3. In https://github.com/oauth-xx/oauth2/commit/7e5cd6c230db50ea981f965e4e95b4f3f190efd5 adds additional parsing for the tokens, and this has to be reconciled now with `extract_access_token` and `access_token_class` behavior.
